atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/monitoring/TimestampTracker.java

```diff
@@ -0,0 +1,96 @@
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.monitoring;
+
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.CachedGauge;
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.Gauge;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.cleaner.Cleaner;
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.logsafe.SafeArg;
+
+public class TimestampTracker implements AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(TimestampTracker.class);
+
+    // We cache underlying calls, in case a hyper-aggressive metrics client repeatedly queries the values.
+    @VisibleForTesting
+    static final Duration CACHE_INTERVAL = Duration.ofSeconds(10L);
+
+    private final MetricsManager metricsManager = new MetricsManager();
+    private final Set<String> registeredMetricShortNames = Sets.newConcurrentHashSet();
+
+    private final Clock clock;
+
+    public TimestampTracker(Clock clock) {
+        this.clock = clock;
+    }
+
+    public static TimestampTracker createWithDefaultTrackers(TimelockService timeLockService, Cleaner cleaner) {
+        TimestampTracker tracker = new TimestampTracker(Clock.defaultClock());
+        tracker.registerTimestampForTracking("timestamp.fresh", timeLockService::getFreshTimestamp);
+        tracker.registerTimestampForTracking("timestamp.immutable", timeLockService::getImmutableTimestamp);
+        tracker.registerTimestampForTracking("timestamp.unreadable", cleaner::getUnreadableTimestamp);
+        return tracker;
+    }
+
+    @VisibleForTesting
+    void registerTimestampForTracking(String shortName, Supplier<Long> supplier) {
+        if (!putMetricIfAbsent(shortName)) {
+            log.warn("A metric with the name {} has already been registered in the timestamp tracker",
+                    SafeArg.of("metricName", shortName));
+            throw new IllegalStateException("A metric with the name " + shortName + " has already been registered!");
+        }
+        metricsManager.registerMetric(TimestampTracker.class, shortName, createCachingGauge(supplier));
+    }
+
+    private boolean putMetricIfAbsent(String shortName) {
+        synchronized (registeredMetricShortNames) {
+            if (registeredMetricShortNames.contains(shortName)) {
+                return false;
+            }
+            registeredMetricShortNames.add(shortName);
+            return true;
+        }
+    }
+
+    private <T> Gauge<T> createCachingGauge(Supplier<T> supplier) {
+        return new CachedGauge<T>(clock, CACHE_INTERVAL.getSeconds(), TimeUnit.SECONDS) {
+            @Override
+            protected T loadValue() {
+                return supplier.get();
```

Comment:
do we know what happens if this supplier throws?

Reply:
Ah, good call. We will get an error logged on the `ScheduledReporter` that's used to grab the metrics. I think it's probably preferable here to catch `Exception` and return the last value in that case. Will change that